pkg/utils/apply/share.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apply
+
+import (
+	"strings"
+
+	"k8s.io/utils/strings/slices"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+)
+
+const (
+	sharedBySep = ","
+)
+
+// AddSharer add sharer
+func AddSharer(sharedBy string, app *v1beta1.Application) string {
```

Comment:
How to guarantee non-conflict? If multiple applications are published at the same time

Reply:
Conflict will be rejected by Kubernetes APIServer. For example, if the resource has sharer A now, B & C both wants to add themselves to the sharer annotation. B&C sends the patch request at the same time.

B&C requests will be handled by kube-apiserver in some order, not concurrently. Let's say B is handled first, the sharer will become "A,B". C's request then will be rejected by kube-apiserver since it is trying to mutate based on an old resourceVersion for that resource. Then it will fail the reconcile and try again. Next time he will find sharer "A,B" at first, and then try to add C to it. The conflicts will be solved in this way.